### ISSUE TITLE
As a user, reserve capacity - add payment flow 

### ISSUE THREAD

**Issue by gmachtel** [threefoldtech/jumpscaleX_threebot#138]
Extension of 

**More info**
Part of the [Container Deployment flow](https://docs.google.com/drawings/d/1KoBF_cYLPE5YoFM85bxsWEl_RlJEoA16pMlFUZDPhSk/edit?usp=sharing)
[Process flow](https://github.com/threefoldtech/tf_cockpit/blob/master/specs/architecture/ProcessFlow.md)

Before implementation; please add to this issue:
- Actor in and outputs
- How does this request from the node end up in the cockpit; what is the technical flow

**Mockups**

![Reservation](https://github.com/threefoldtech/jumpscaleX_threebot/raw/development/docs/images/GridUserReserveCapacity3.png) 
![NodeReservationCalculus](https://github.com/threefoldtech/jumpscaleX_threebot/raw/development/docs/images/GridUserReserveCapacity%20Kart2.png) 
![network_reservation_from](https://user-images.githubusercontent.com/303164/67089605-5b4bba00-f1a8-11e9-8617-ec1b9591748e.png)
![container_reservation_from](https://user-images.githubusercontent.com/303164/67089606-5be45080-f1a8-11e9-97a8-c508a06c6c6c.png)
![0-db_reservation_from](https://user-images.githubusercontent.com/303164/67089607-5be45080-f1a8-11e9-8f4b-18789bef4989.png)

**DOD**
- Actors are working and tested
- UI is linked to actors
- Node or nodes chosen (after search based on available capacity, location, certified capacity Y/N)
- Capacity chosen for nodes (CRU, MRU, HRU, SRU) 
- Reservation duration chosen
- Wireguard configuration filled
- Result of action is 'capacity reservation pending with farmer to be signed'
- On join notification is received both with consumer and farmer

**Open questions**
- Pay before confirmation by farmer ? If direct payment and farmer refuses, refund needed. Other possibility: lock the funds.

**Comment by robvanmieghem** [comment]
I think an option is to provide the farmer with an unpublished transaction that he has to co-sign and publish to claim the funds or something similar on-chain where the farmer can claim the funds until a specific timeout expires, in which case the requester can reclaim them. The on-chain solution provides protection against double spending.

**Comment by gmachtel** [comment]
I think on-chain solution is acceptable now as there is no time criticality at the moment. Locking of funds seems more important at this stage, I believe.

**Comment by robvanmieghem** [comment]
My idea was just an initial one, we might come up with a better on-chain contract. @GlenDC your thoughts?

**Comment by GlenDC** [comment]
I do not think we have to reinvent the wheel here. There are plenty of other similar solutions that require such a contract. I suggest to look at those and see what we can come up with.

This discussion is not new though, and I do not think I have new info to add since we talked about this many times before. When a user reserves capacity for a given period (dunno what our rent periods are? That average month, or what?) than it can be assumed that the user might run critical software on this (critical for that user) and expects the software to run perfectly fine for that entire period, anything less is unacceptable. Therefore you do want to lock up the money indeed for that entire period. There are several I see of handling a contract that breaks prior to the end date:

- either you do pay out the farmer for the time served (doesn't have to be linear, can also be done in a way that on top of that a penalty is subtracted and destroyed, where the earlier it breaks the bigger the penalty is, a penalty which itself also does not have to be linear, and probably shouldn't be), with part of the money returned to the user
- either the farmer gets nothing when the contract breaks, and the user neither (this to prevent providing incentive to the user of trying to sabotage itself, to get all  the money back while having enjoyed some of the service already)
- we go for option (1), but the user gets nothing back, to prevent incentive that the user might actually want a short contract (while this might not be available) and somehow ensures the server stops working to pay for the desired time only)

Perhaps there is even an entire different road. If contracts are also possible for very long durations (anything more than a month I would say) you might need to work something out that divides the contract in periods first, so that the farmer does not have to wait too long for the money, and that we look at breaking of the contract without it having impact on previously served periods of the same contract.

From previous discussions it was concluded that while the details of the contracts would have to be worked out, the difficulty lays not here. The difficulty will lay in providing a (neutral?) proof of whether or not the contract was broken.

Also given that the payments are done on chain, it makes sense the contract lives on the chain as well. That does however not mean that there is a need for interaction with the chain other than creating and claiming a contract. Perhaps even the claiming can be done automatically. The proving for example can probably be completely done off-chain with the created contract to provide the expectations. Once a break of contract has been proven, it does somehow have to be linked to the contract.

However, you could also provide broker bots that keep money locked up in their own pool, and do the entire smart contracts off-chain, such that from a chain perspective it is simply a transfer of coins from many to one, and one to many accounts.

Given that we're dealing with real money I do not think we should cut corners here and do anything less than what is proposed. I cannot think of an in-between solution that is acceptable. We still have until beginning of next year to get this right though, so once we align on how this should be done, we could make it our top priority to get this right from all different teams.

**Comment by GlenDC** [comment]
Also can somebody explain what this phonebook is and how it plays in the story. When I think of a phonebook I think of a simply dictionary that links contact information with personal details, as an easy way for humans to look up someone's contact information. However, given that I've seen (by accident) in the past that within the JSX codebase the phonebook seems to also create and manage entire TF wallets, I would like to get a clear description of what the phonebook does, how big its scope is, whether it is private (and known only) to one 3bot, and so on. As the linked picture does not really explain much about it at all.

**Comment by zaibon** [comment]
@GlenDC the phonebook is just a public directory of all the 3bot identities from the grid. It contains entry that represent a 3bot (double name, public key, address).

The phonebook doens't managed people and/or 3bot wallets.

**Comment by GlenDC** [comment]
How does it get all that data (as in where is its input coming from, exactly)? And can anyone build this directory or is this a central piece we provide somehow?

**Comment by gmachtel** [comment]
Phonebook direcory is fed through the registration process : https://github.com/threefoldtech/jumpscaleX_threebot/issues/118

**Comment by GlenDC** [comment]
Ok so that is a centralized phonebook only we can populate? Is there also a long term decentralized solution (or road towards such a solution) we have in mind?

**Comment by gmachtel** [comment]
It's something like the DNS (also there Network Solutions is still the centralizing organisation). So full decentralization won't be possible I think, but it shows that the number of fields in the list should be restricted to the bare minimum: 3bot double name, IP address, wallet public address, and that's it.
